File: artifacts/api-server/src/lib/trace-diff.ts
/*
 * Trace v2 diff — TS mirror of TraceDiff.cpp. Both implementations
 * read the same byte-level format and compare frame-by-frame in
 * lock-step. Determinism is the headline guarantee, so this diff is
 * structural: every divergence is reported as one of a small set of
 * stable kinds.
 */

import { parseTraceV2, type TraceFrame, type EntityFrame } from "./trace-reader";
import type { TraceDiffEntry } from "@workspace/api-zod/types/traceDiffEntry";

export type { TraceDiffEntry };

export interface TraceDiffOutcome {
  identical: boolean;
  firstDivergenceTick?: number;
  entries: TraceDiffEntry[];
}

const MAX_ENTRIES = 100;

export function diffTraces(a: Buffer, b: Buffer): TraceDiffOutcome {
  const out: TraceDiffOutcome = { identical: true, entries: [] };

  let parsedA: ReturnType<typeof parseTraceV2>;
  let parsedB: ReturnType<typeof parseTraceV2>;
  try {
    parsedA = parseTraceV2(a);
  } catch (err) {
    return note(out, 0, "frame_size_diff",
      `trace A failed to load: ${(err as Error).message}`);
  }
  try {
    parsedB = parseTraceV2(b);
  } catch (err) {
    return note(out, 0, "frame_size_diff",
      `trace B failed to load: ${(err as Error).message}`);
  }

  const fa = parsedA.frames;
  const fb = parsedB.frames;
  const common = Math.min(fa.length, fb.length);

  for (let i = 0; i < common; i++) {
    const frameA = fa[i]!;
    const frameB = fb[i]!;
    if (frameA.tick !== frameB.tick) {
      note(out, frameA.tick, "frame_size_diff",
        `tick mismatch: A=${frameA.tick} B=${frameB.tick}`);
      continue;
    }
    diffEntities(out, frameA, frameB);
    diffEvents(out, frameA, frameB);
  }

  if (fa.length !== fb.length) {
    note(out, common, "frame_size_diff",
      `frame count differs: A=${fa.length} B=${fb.length}`);
  }

  return out;
}

function diffEntities(out: TraceDiffOutcome, a: TraceFrame, b: TraceFrame): void {
  const ne = Math.min(a.entities.length, b.entities.length);
  for (let e = 0; e < ne; e++) {
    const ea = a.entities[e]!;
    const eb = b.entities[e]!;
    if (ea.index !== eb.index || ea.generation !== eb.generation) {
      note(out, a.tick, "entity_changed",
        `entity slot ${e} identity differs (A=${ea.index}/${ea.generation} ` +
        `B=${eb.index}/${eb.generation})`);
      continue;
    }
    if (!entitiesEqual(ea, eb)) {
      note(out, a.tick, "entity_changed",
        `entity ${ea.index} components differ`);
    }
  }
  for (let e = ne; e < a.entities.length; e++) {
    note(out, a.tick, "entity_removed",
      `entity ${a.entities[e]!.index} present in A only`);
  }
  for (let e = ne; e < b.entities.length; e++) {
    note(out, a.tick, "entity_added",
      `entity ${b.entities[e]!.index} present in B only`);
  }
}

function diffEvents(out: TraceDiffOutcome, a: TraceFrame, b: TraceFrame): void {
  const ev = Math.min(a.events.length, b.events.length);
  for (let k = 0; k < ev; k++) {
    if (a.events[k] !== b.events[k]) {
      note(out, a.tick, "event_diff",
        `event ${k} differs: A='${a.events[k]}' B='${b.events[k]}'`);
    }
  }
  if (a.events.length !== b.events.length) {
    note(out, a.tick, "event_diff",
      `event count differs: A=${a.events.length} B=${b.events.length}`);
  }
}

function entitiesEqual(a: EntityFrame, b: EntityFrame): boolean {
  if (a.pacId !== b.pacId) return false;
  if (a.type !== b.type) return false;
  const pa = a.position;
  const pb = b.position;
  if ((pa === undefined) !== (pb === undefined)) return false;
  if (pa && pb) {
    if (pa.x !== pb.x || pa.y !== pb.y || pa.z !== pb.z) return false;
  }
  return true;
}

function note(
  out: TraceDiffOutcome,
  tick: number,
  kind: TraceDiffEntry["kind"],
  detail: string,
): TraceDiffOutcome {
  out.identical = false;
  if (out.firstDivergenceTick === undefined) {
    out.firstDivergenceTick = tick;
  }
  if (out.entries.length < MAX_ENTRIES) {
    out.entries.push({ tick, kind, detail });
  }
  return out;
}

File: pacengine/engine/trace/src/TraceDiff.cpp
#include "TraceDiff.hpp"

#include "TraceReader.hpp"

#include <algorithm>
#include <sstream>

namespace pac {

namespace {

void note(TraceDiffResult& out,
          std::uint64_t tick,
          const std::string& kind,
          const std::string& detail,
          std::size_t max_entries) {
    out.identical = false;
    if (!out.first_divergence_tick.has_value()) {
        out.first_divergence_tick = tick;
    }
    if (out.entries.size() < max_entries) {
        out.entries.push_back({tick, kind, detail});
    }
}

bool components_equal(const TraceEntity& a, const TraceEntity& b) {
    if (a.components.size() != b.components.size()) return false;
    for (std::size_t i = 0; i < a.components.size(); ++i) {
        if (a.components[i].type_tag != b.components[i].type_tag) return false;
        if (a.components[i].payload  != b.components[i].payload)  return false;
    }
    return true;
}

} // namespace

TraceDiffResult TraceDiff::diff(const std::string& path_a,
                                const std::string& path_b,
                                std::size_t max_entries) {
    TraceDiffResult out;

    TraceReader ra(path_a);
    TraceReader rb(path_b);
    if (!ra.ok()) {
        note(out, 0, "frame_size_diff",
             "trace A failed to load: " + ra.error(), max_entries);
        return out;
    }
    if (!rb.ok()) {
        note(out, 0, "frame_size_diff",
             "trace B failed to load: " + rb.error(), max_entries);
        return out;
    }

    const auto& fa = ra.frames();
    const auto& fb = rb.frames();
    const std::size_t common = std::min(fa.size(), fb.size());

    for (std::size_t i = 0; i < common; ++i) {
        const TraceFrame& a = fa[i];
        const TraceFrame& b = fb[i];

        if (a.tick != b.tick) {
            std::ostringstream d;
            d << "tick mismatch: A=" << a.tick << " B=" << b.tick;
            note(out, a.tick, "frame_size_diff", d.str(), max_entries);
            continue;
        }

        // Entity-level diff. Match by (index, generation) — same key the
        // writer uses, so this is the deterministic identity comparison.
        // Entities are written in stable document order, so we can walk
        // both lists in lock-step.
        const std::size_t ne = std::min(a.entities.size(), b.entities.size());
        for (std::size_t e = 0; e < ne; ++e) {
            const auto& ea = a.entities[e];
            const auto& eb = b.entities[e];
            if (ea.index != eb.index || ea.generation != eb.generation) {
                std::ostringstream d;
                d << "entity slot " << e << " identity differs (A="
                  << ea.index << "/" << ea.generation
                  << " B=" << eb.index << "/" << eb.generation << ")";
                note(out, a.tick, "entity_changed", d.str(), max_entries);
                continue;
            }
            if (!components_equal(ea, eb)) {
                std::ostringstream d;
                d << "entity " << ea.index << " components differ";
                note(out, a.tick, "entity_changed", d.str(), max_entries);
            }
        }
        for (std::size_t e = ne; e < a.entities.size(); ++e) {
            std::ostringstream d;
            d << "entity " << a.entities[e].index << " present in A only";
            note(out, a.tick, "entity_removed", d.str(), max_entries);
        }
        for (std::size_t e = ne; e < b.entities.size(); ++e) {
            std::ostringstream d;
            d << "entity " << b.entities[e].index << " present in B only";
            note(out, a.tick, "entity_added", d.str(), max_entries);
        }

        // Events. Compare in order; mismatch ⇒ event_diff.
        const std::size_t ev = std::min(a.events.size(), b.events.size());
        for (std::size_t k = 0; k < ev; ++k) {
            if (a.events[k] != b.events[k]) {
                std::ostringstream d;
                d << "event " << k << " differs: A='" << a.events[k]
                  << "' B='" << b.events[k] << "'";
                note(out, a.tick, "event_diff", d.str(), max_entries);
            }
        }
        if (a.events.size() != b.events.size()) {
            std::ostringstream d;
            d << "event count differs: A=" << a.events.size()
              << " B=" << b.events.size();
            note(out, a.tick, "event_diff", d.str(), max_entries);
        }
    }

    if (fa.size() != fb.size()) {
        std::ostringstream d;
        d << "frame count differs: A=" << fa.size() << " B=" << fb.size();
        note(out, common, "frame_size_diff", d.str(), max_entries);
    }

    return out;
}

} // namespace pac
